src/dialogs/more-info/controls/more-info-lock.ts

```diff
@@ -1,43 +1,156 @@
-import "@material/mwc-button";
-import type { HassEntity } from "home-assistant-js-websocket";
-import { css, html, LitElement, nothing } from "lit";
-import { customElement, property, query } from "lit/decorators";
+import "@material/web/iconbutton/outlined-icon-button";
+import { mdiDoorOpen, mdiLock, mdiLockOff } from "@mdi/js";
+import { css, CSSResultGroup, html, LitElement, nothing } from "lit";
+import { customElement, property } from "lit/decorators";
+import { styleMap } from "lit/directives/style-map";
+import { domainIcon } from "../../../common/entity/domain_icon";
+import { stateColorCss } from "../../../common/entity/state_color";
+import { supportsFeature } from "../../../common/entity/supports-feature";
 import "../../../components/ha-attributes";
-import "../../../components/ha-textfield";
-import type { HaTextField } from "../../../components/ha-textfield";
+import { UNAVAILABLE } from "../../../data/entity";
+import { LockEntity, LockEntityFeature } from "../../../data/lock";
 import type { HomeAssistant } from "../../../types";
+import { showEnterCodeDialogDialog } from "../../enter-code/show-enter-code-dialog";
+import { moreInfoControlStyle } from "../components/ha-more-info-control-style";
+import "../components/ha-more-info-state-header";
+import "../components/lock/ha-more-info-lock-toggle";
 
 @customElement("more-info-lock")
 class MoreInfoLock extends LitElement {
   @property({ attribute: false }) public hass!: HomeAssistant;
 
-  @property({ attribute: false }) public stateObj?: HassEntity;
+  @property({ attribute: false }) public stateObj?: LockEntity;
 
-  @query("ha-textfield") private _textfield?: HaTextField;
+  private async _open() {
+    this._callService("open");
+  }
+
+  private async _lock() {
+    this._callService("lock");
+  }
+
+  private async _unlock() {
+    this._callService("unlock");
+  }
+
+  private async _callService(service: "open" | "lock" | "unlock") {
+    let code: string | undefined;
+
+    if (this.stateObj!.attributes.code_format) {
+      const response = await showEnterCodeDialogDialog(this, {
+        codeFormat: "text",
+        codePattern: this.stateObj!.attributes.code_format,
+        title: this.hass.localize(
+          `ui.dialogs.more_info_control.lock.${service}`
+        ),
+        submitText: this.hass.localize(
+          `ui.dialogs.more_info_control.lock.${service}`
+        ),
+      });
+      if (!response) {
+        return;
+      }
+      code = response;
+    }
+
+    this.hass.callService("lock", service, {
+      entity_id: this.stateObj!.entity_id,
+      code,
+    });
+  }
```

Comment:
We have this code twice (here and `ha-more-info-lock-toggle`), can we share it?

Reply:
We have the same for alarm_control_panel. I will do a another PR to update that